homestar-runtime/src/settings.rs

```diff
@@ -22,6 +31,126 @@ pub struct Monitoring {
+    GenerateFromSeed(PupkeyRNGSeed),
+    /// File path to a PEM encoded ed25519 key
+    #[serde(rename = "existing")]
+    Existing(ExistingKeyPath),
+}
+
+/// Supported key types of homestar
+#[derive(Clone, Debug, Default, Deserialize)]
+pub(crate) enum KeyType {
+    #[default]
+    #[serde(rename = "ed25519")]
+    Ed25519,
+    #[serde(rename = "secp256k1")]
+    Secp256k1,
+}
+
+/// Seed material for RNG generated keys
+#[serde_as]
+#[derive(Clone, Debug, Deserialize)]
+pub(crate) struct PupkeyRNGSeed {
+    #[serde(default)]
+    key_type: KeyType,
+    #[serde_as(as = "Base64")]
+    seed: [u8; 32],
+}
+
+/// Info on where and what the Key file is
+#[derive(Clone, Debug, Deserialize)]
+pub(crate) struct ExistingKeyPath {
+    #[serde(default)]
+    key_type: KeyType,
+    path: String,
+}
+
+impl PubkeyConfig {
+    /// Produce a Keypair using the given configuration.
+    /// Calling this function will access the filesystem if configured to import a key.
+    pub(crate) fn keypair(&self) -> anyhow::Result<identity::Keypair> {
+        match self {
+            PubkeyConfig::Random => {
+                info!("generating random ed25519 key");
+                Ok(identity::Keypair::generate_ed25519())
+            }
+            PubkeyConfig::GenerateFromSeed(PupkeyRNGSeed { key_type, seed }) => {
+                // seed RNG with supplied seed
+                let mut r = rand::prelude::StdRng::from_seed(*seed);
+                let mut new_key: [u8; 32] = r.gen();
+
+                match key_type {
+                    KeyType::Ed25519 => {
+                        info!("generating radom ed25519 key from seed");
+
+                        identity::Keypair::ed25519_from_bytes(new_key).map_err(|e| {
+                            anyhow!("failed to generate ed25519 key from random: {:?}", e)
+                        })
+                    }
+                    KeyType::Secp256k1 => {
+                        info!("generating radom secp256k1 key from seed");
+
+                        let sk =
+                            secp256k1::SecretKey::try_from_bytes(&mut new_key).map_err(|e| {
+                                anyhow!("failed to generate secp256k1 key from random: {:?}", e)
+                            })?;
+                        let kp = secp256k1::Keypair::from(sk);
+                        Ok(identity::Keypair::from(kp))
+                    }
+                }
+            }
+            PubkeyConfig::Existing(ExistingKeyPath { key_type, path }) => {
+                let path = Path::new(&path);
+                let mut file = std::fs::File::open(path).context("unable to read key file")?;
+
+                let mut buf = Vec::new();
+                file.read_to_end(&mut buf)
+                    .context("unable to read bytes from file, is the file corrupted?")?;
+
+                match key_type {
+                    KeyType::Ed25519 => {
+                        const PEM_HEADER: &str = "PRIVATE KEY";
```

Comment:
minor: I tend to put consts up top, though doesn't matter much.

Reply:
Didn't want it conflated when we have the secp PEM headers, so would pref to keep as is.